parquet-column/src/main/java/org/apache/parquet/schema/Types.java

```diff
@@ -555,114 +622,1267 @@ protected GroupType build(String name) {
+        return new ModMapBuilder<P>((MapBuilder<P>)(
+            (BaseMapBuilder.MapListKeyBuilder) parent).parent).repetition(repetition);
+      }
+
+      public ModMapBuilder<P> requiredMapValue() {
+        return mapValue(Type.Repetition.REQUIRED);
+      }
+
+      public ModMapBuilder<P> optionalMapValue() {
+        return mapValue(Type.Repetition.OPTIONAL);
+      }
+
+      public class ModListBuilder<I> extends BaseListBuilder<I,
+          ModListBuilder<I>> {
+
+        MapBuilder<I> parent;
+
+        public ModListBuilder(MapBuilder<I> parent) {
+          super(parent.parent);
+          this.parent = parent;
+        }
+
+        @Override
+        protected ModListBuilder<I> self() {
+          return this;
+        }
+
+        @Override
+        public I named(String name) {
+          parent.setValueType(build("value"));
+          return parent.named(name);
+        }
+      }
+
+      public ModListBuilder<P> listValue(Type.Repetition repetition) {
+        parent.setElementType(super.build("element"));
+        parent.repetition(Type.Repetition.REQUIRED);
+        ((BaseMapBuilder.MapListKeyBuilder) parent).parent.setKeyType(parent.build("key"));
+        return new ModListBuilder<P>((MapBuilder<P>)(
+            (BaseMapBuilder.MapListKeyBuilder) parent).parent).repetition(repetition);
+      }
+
+      public ModListBuilder<P> requiredListValue() {
+        return listValue(Type.Repetition.REQUIRED);
+      }
+
+      public ModListBuilder<P> optionalListValue() {
+        return listValue(Type.Repetition.OPTIONAL);
+      }
+    }
+
+    @Override
+    protected abstract Q self();
+
+    @Override
+    protected Type build(String name) {
+      Preconditions.checkNotNull(elementType, "List element type");
+      return buildGroup(repetition)
+          .as(OriginalType.LIST)
+          .repeatedGroup()
+          .addFields(elementType)
+          .named("list")
+          .named(name);
+    }
+
+    public P named(String name) {
+      Preconditions.checkState(originalType == null,
+          "LIST is already the logical type and can't be changed");
+      return super.named(name);
+    }
+
+    public ListGroupElementBuilder<P, Q> groupElement(Type.Repetition repetition) {
+      Preconditions
+          .checkState(elementType == null, "groupElement can be called just once on a ListBuilder");
+      if (parent != null) {
+        return new ListGroupElementBuilder<P, Q>(this).repetition
+            (repetition);
+      } else {
+        return new ListGroupElementBuilder<P, Q>(this, returnType).repetition
+            (repetition);
+      }
+    }
+
+    public ListGroupElementBuilder<P, Q> requiredGroupElement() {
+      return groupElement(Type.Repetition.REQUIRED);
+    }
+
+    public ListGroupElementBuilder<P, Q> optionalGroupElement() {
+      return groupElement(Type.Repetition.OPTIONAL);
+    }
+
+    public ListElementBuilder<P, Q> element(PrimitiveTypeName type, Type.Repetition repetition) {
```

Comment:
You still need to add the versions that take types: `element(Type)`, `requiredElement(Type)`, and `optionalElement(Type)`